src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+    sperror("READ", drive->err);
+    return (drive->err);
+  }
+  return 0;
+}
+
+int read_one_ecc_block(drive_info *drive, unsigned char *data, int32_t lba) {
+  drive->cmd[0] = MMC_READ;
+  drive->cmd[2] = (lba >> 24) & 0xFF;
+  drive->cmd[3] = (lba >> 16) & 0xFF;
+  drive->cmd[4] = (lba >> 8) & 0xFF;
+  drive->cmd[5] = lba & 0xFF;
+  drive->cmd[8] = 0x10;
+  if ((drive->err = drive->cmd.transport(READ, data, 0x8000))) {
+    sperror("READ_ONE_ECC_BLOCK", drive->err);
+    return (drive->err);
+  }
+  //	if ((drive->err=drive->cmd.transport(READ,(void*)((unsigned
+  // char*)drive->rd_buf+(0x0001<<14)),0x34))) 		{sperror
+  //("READ_ONE_ECC_BLOCK",drive->err); return (drive->err);}
+  return 0;
+}
+
+int get_drive_serial_number(drive_info *drive) {
+  //	char data[2048]; memset(data, 0, sizeof(data));
+  unsigned int data_length;
+  unsigned int length;
+  get_configuration(drive, FEATURE_LOGICAL_UNIT_SERIAL_NUMBER, &data_length,
+                    NULL);
+  if (drive->err)
+    return -1;
+  length = drive->rd_buf[11];
+  drive->rd_buf[12 + length] = 0;
+  if (data_length > 8)
+    strncpy(drive->serial, (char *)drive->rd_buf + 12, 16);
+  else
+    drive->serial[0] = 0;
+  return 0;
+}
+
+int get_buffer_capacity(drive_info *drive) {
+  int offs;
+  if (mode_sense(drive, MODE_PAGE_MM_CAP_STATUS, 0, 192))
+    return 1;
+  offs = 0;
+  while (((drive->rd_buf[offs]) & 0x3F) != 0x2A)
+    offs++;
+  drive->buffer_size = ntoh16u(drive->rd_buf + offs + 0x0C);
+  //	printf("Buffer capacity: 0x%04X (%d)KB\n", drive->buffer_size,
+  // drive->buffer_size);
+  return 0;
+}
+
+int get_wbuffer_capacity(drive_info *drive, uint32_t *btot, uint32_t *bfree) {
+  uint8_t td[0x0C];
+  memset(td, 0, 0x0C);
+  drive->cmd[0] = MMC_READ_BUFFER_CAPACITY;
+  drive->cmd[8] = 0x0C;
+  if ((drive->err = drive->cmd.transport(READ, td, 0x0C))) {
+    sperror("READ_BUFFER_CAPACITY", drive->err);
+    return (drive->err);
+  }
+
+  (*btot) = ntoh32u(td + 4);
+  (*bfree) = ntoh32u(td + 8);
+  return 0;
+}
+
+int read_writer_info(drive_info *drive) {
+  if (!(drive->media.type & DISC_DVDminus)) {
+    strcpy(drive->media.writer, "n/a (only for DVD-R(W))");
+    return 1;
+  }
+  char format = 0x0D;
+  drive->media.writer[0] = 0;
+  drive->rd_buf[8] = 0;
+  drive->cmd[0] = MMC_READ_DVD_STRUCTURE;
+  drive->cmd[5] = 0x02;
+  drive->cmd[7] = format;
+  drive->cmd[8] = 8;
+  drive->cmd[9] = 8;
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, 2056)) ||
+      (!drive->rd_buf[8])) {
+    printf("Read Writer Info failed\n");
+    return 1;
+  }
+  for (int i = 0; i < 0x3F; i++) {
+    if (!drive->rd_buf[8 + i])
+      drive->rd_buf[8 + i] = 0x20;
+  }
+  strncpy(drive->media.writer, (char *)drive->rd_buf + 8, 0x3F);
+  remove_double_spaces(drive->media.writer);
+  //	remove_end_spaces(drive->media.writer);
+  return 0;
+}
+
+int detect_mm_capabilities(drive_info *drive) {
+  char len;
+  int offs;
+  int i, j;
```

Comment:
Codacy found an issue: [The scope of the variable 'i' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)